src/foundry/client/data/documents/chat-message.d.mts

```diff
@@ -486,31 +493,95 @@ declare global {
+      /** The Scene in which the speaker resides */
+      scene: Scene.Implementation;
 
       /** The name of the speaker to display */
-      alias: string | undefined;
+      alias: string;
+    }
+
+    interface GetSpeakerOptions extends NullishProps<_BaseSpeakerOptions> {}
+
+    /** @remarks `token` is required, `alias` falls back to `token.name` */
+    interface GetSpeakerFromTokenOptions extends NullishProps<Pick<_BaseSpeakerOptions, "token" | "alias">, "alias"> {}
+
+    /** @remarks `actor` is required, `scene` falls back to `canvas.scene`, `alias` falls back to `actor.name` */
+    interface GetSpeakerFromActorOptions
+      extends NullishProps<Pick<_BaseSpeakerOptions, "scene" | "actor" | "alias">, "scene" | "alias"> {}
+
+    /** @remarks `user` is required, `scene` falls back to `canvas.scene`, `alias` falls back to `user.name` */
+    interface GetSpeakerFromUserOptions extends NullishProps<Pick<_BaseSpeakerOptions, "scene" | "alias">> {
+      /** The User who is speaking */
+      user: User.Implementation;
     }
 
+    /** @internal */
+    type _SpeakerData = SchemaField.InitializedData<ChatMessage.Schema>["speaker"];
+
+    interface SpeakerData extends _SpeakerData {}
+
+    /**
+     * @remarks
+     * {@linkcode ChatMessage.getWhisperRecipients} has a couple special-cased values, and a couple fallback behaviors.
+     * _ALL_ comparisons are case-**in**sensitive, compared lowercase.
+     * - `"GM"` or `"DM"` inputs returns `game.users.filter(u => u.isGM)`
+     * - `"players"` returns `game.users.players`
+     * - Then if any User names match, returns all that do
+     * - Then returns any Users whose assigned `character` matches
+     * - Finally returns `[]`
+     */
+    type WhisperRecipient = "GM" | "DM" | "players" | (string & {});
+
+    /**
+     * @remarks Serves two purposes:
+     * - Template context for either calling `renderTemplate` on `CONFIG.ChatMessage.template` or passing to {@link ChatMessage._renderRollContent | `ChatMessage#_renderRollContent`}
+     * - Context passed to the {@linkcode Hooks.StaticCallbacks.renderChatMessage | `renderChatMessage`} hook.
+     */
     interface MessageData {
+      /** @remarks This is the return of `ChatMessage#toObject(false)`, but that doesn't seem to make a difference */
       message: ChatMessage.Source;
+
+      /** @remarks Always `game.user` */
       user: User.Stored;
-      author: User.Implementation | undefined;
+
+      /** @remarks The message's {@link ChatMessage.author | `author`} */
+      author: User.Implementation;
+
+      /** @remarks The message's {@link ChatMessage.alias | `alias`} */
       alias: string;
+
+      /** @remarks Possibly more than one class name, space-separated */
       cssClass: string;
-      isWhisper: boolean;
+
+      /**
+       * @remarks The `.length` of the message's {@link ChatMessage.whisper | `whisper`} array,
+       * despite the name implying a `boolean`
+       */
+      isWhisper: number;
```

Comment:
That's... rather odd. Do you think it should be `!!this.whisper.length`? Does Foundry document this as `boolean`? If so that's a documentation issue.